test/domain_bridge/test_domain_bridge.cpp

```diff
@@ -0,0 +1,179 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <string>
+
+#include "rclcpp/executors/single_threaded_executor.hpp"
+#include "rclcpp/context.hpp"
+
+#include "domain_bridge/domain_bridge.hpp"
+
+class TestDomainBridge : public ::testing::Test
+{
+protected:
+  static void SetUpTestCase()
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  static void TearDownTestCase()
+  {
+    rclcpp::shutdown();
+  }
+};
+
+TEST_F(TestDomainBridge, construction_destruction)
+{
+  // Default
+  {
+    domain_bridge::DomainBridge bridge;
+  }
+  // With options
+  {
+    domain_bridge::DomainBridge bridge(domain_bridge::DomainBridgeOptions());
+  }
+}
+
+TEST_F(TestDomainBridge, get_options)
+{
+  domain_bridge::DomainBridgeOptions input_options;
+  input_options.name("domain_bridge_test_name");
+  domain_bridge::DomainBridge bridge(input_options);
+  domain_bridge::DomainBridgeOptions output_options = bridge.get_domain_bridge_options();
+  EXPECT_EQ(input_options.name(), output_options.name());
+}
+
+TEST_F(TestDomainBridge, bridge_topic_valid)
+{
+  // Nominal test
+  {
+    domain_bridge::DomainBridge bridge;
+    bridge.bridge_topic("foo", "test_msgs/msg/BasicTypes", 1, 2);
+  }
+  // Same domain ID
+  {
+    domain_bridge::DomainBridge bridge;
+    bridge.bridge_topic("foo", "test_msgs/msg/BasicTypes", 1, 1);
+  }
+  // Topic with namespace
+  {
+    domain_bridge::DomainBridge bridge;
+    bridge.bridge_topic("foo/bar/baz", "test_msgs/msg/BasicTypes", 1, 2);
+  }
+  // Multiple topics
+  {
+    domain_bridge::DomainBridge bridge;
+    bridge.bridge_topic("foo", "test_msgs/msg/BasicTypes", 1, 2);
+    bridge.bridge_topic("bar", "test_msgs/msg/BasicTypes", 1, 2);
+    bridge.bridge_topic("baz", "test_msgs/msg/BasicTypes", 1, 2);
+  }
+  // Same topic, different domains
+  {
+    domain_bridge::DomainBridge bridge;
+    bridge.bridge_topic("foo", "test_msgs/msg/BasicTypes", 1, 2);
+    bridge.bridge_topic("foo", "test_msgs/msg/BasicTypes", 1, 3);
+    bridge.bridge_topic("foo", "test_msgs/msg/BasicTypes", 2, 1);
+  }
+  // Same topic, different types
+  // This use-case isn't clearly supported by ROS 2, see https://github.com/ros2/ros2/issues/1095
+  // {
+  //   domain_bridge::DomainBridge bridge;
+  //   bridge.bridge_topic("foo", "test_msgs/msg/BasicTypes", 1, 2);
+  //   bridge.bridge_topic("foo", "test_msgs/msg/Nested", 1, 2);
+  //   bridge.bridge_topic("foo", "test_msgs/msg/Constants", 1, 2);
+  // }
+}
+
+TEST_F(TestDomainBridge, bridge_topic_invalid)
+{
+  // Invalid topic name
+  {
+    domain_bridge::DomainBridge bridge;
+    EXPECT_THROW(
+      bridge.bridge_topic("Not a v@lid topic name!", "test_msgs/msg/BasicTypes", 1, 2),
+      rclcpp::exceptions::InvalidTopicNameError);
```

Comment:
I'd like to catch and throw exceptions defined in the `domain_bridge` namespace, for all invalid cases here that are throwing.
I'll do that in a separate PR if that sounds good to others.

Reply:
That's an interesting question, in general. The benefits of doing it that way include having a nice, clean abstract interface which includes domain_bridge-specific exceptions (only). However, with C++, if one relies on other libraries in the implementation, it's very hard to know which exceptions might be thrown. So, there's no guarantee to the user of this class that the provided exceptions are the only ones that might be thrown (unless you do something like catch (...)). Also, another drawback is having to catch exceptions in the implementation and then throwing semantically equivalent exceptions in the domain bridge namespace. I've typically approached catching exceptions with the idea that one should catch an exception if the code at that point can do something about the exception, otherwise, let it pass and maybe someone else along the call stack will handle it. So, I'm kind of ambivalent about which way to go. I lean towards your recommendation, but if we go that way, I think we should do this consistently with other ROS2 modules.

Reply:
Good points. My initial thought was to provide a cleaner interface as you describe, but it's true it's difficult to account for all possible exceptions thrown by the underlying libraries. At the very least, I think we can make the interface a bit tidier for common/obvious error cases (like invalid topic or type names).

Reply:
Yes, agreed.

Reply:
IMHO, not catching these `rclcpp` exceptions is ok, they have pretty clear meaning already